server/src/main/scala/services/crunch/deskrecs/DynamicRunnableDeskRecs.scala

```diff
@@ -0,0 +1,152 @@
+import services.SDate
+import services.crunch.desklimits.TerminalDeskLimitsLike
+
+import scala.collection.immutable.Map
+import scala.concurrent.{ExecutionContext, Future}
+
+object DynamicRunnableDeskRecs {
+  val log: Logger = LoggerFactory.getLogger(getClass)
+
+  def createGraph(deskRecsSinkActor: ActorRef,
+                  daysSourceQueue: Source[MillisSinceEpoch, NotUsed],
+                  daysToDeskRecs: DaysToDeskRecs)
+                 (implicit ec: ExecutionContext): RunnableGraph[NotUsed] = {
+
+    val deskRecsSink = Sink.actorRefWithAck(deskRecsSinkActor, StreamInitialized, Ack, StreamCompleted, StreamFailure)
+
+    val graph = GraphDSL.create(daysSourceQueue) {
+      implicit builder =>
+        daysSourceQueueAsync =>
+          daysToDeskRecs(daysSourceQueueAsync.out) ~> deskRecsSink
+          ClosedShape
+    }
+
+    RunnableGraph.fromGraph(graph)
+  }
+
+  type DaysToDeskRecs = Implicits.PortOps[MillisSinceEpoch] => Implicits.PortOps[DeskRecMinutes]
+
+  def daysToDeskRecs(flightsProvider: MillisSinceEpoch => Future[Source[FlightsWithSplits, NotUsed]],
+                     liveManifestsProvider: MillisSinceEpoch => Future[Source[VoyageManifests, NotUsed]],
+                     historicManifestsProvider: Iterable[Arrival] => Future[Map[ArrivalKey, VoyageManifest]],
+                     crunchStart: SDateLike => SDateLike,
+                     splitsCalculator: SplitsCalculator,
+                     portDeskRecs: DesksAndWaitsPortProviderLike,
+                     maxDesksProviders: Map[Terminal, TerminalDeskLimitsLike])
+                    (days: Implicits.PortOps[MillisSinceEpoch])
+                    (implicit ec: ExecutionContext): Implicits.PortOps[DeskRecMinutes] = {
+    val crunchDays = days.map(millis => crunchStart(SDate(millis)).millisSinceEpoch)
+    val withFlights = addFlights(crunchDays, flightsProvider)
+    val withSplits = addSplits(withFlights, liveManifestsProvider, historicManifestsProvider, splitsCalculator)
+    toDeskRecs(withSplits, portDeskRecs, maxDesksProviders)
+  }
+
+  private def toDeskRecs(dayAndFlights: Implicits.PortOps[(MillisSinceEpoch, Iterable[ApiFlightWithSplits])],
+                         portDeskRecs: DesksAndWaitsPortProviderLike,
+                         maxDesksProviders: Map[Terminal, TerminalDeskLimitsLike]
+                        ): Implicits.PortOps[DeskRecMinutes] = {
+    dayAndFlights
+      .map { case (crunchStartMillis, flights) =>
+        val crunchEndMillis = SDate(crunchStartMillis).addMinutes(portDeskRecs.minutesToCrunch).millisSinceEpoch
+        val minuteMillis = crunchStartMillis until crunchEndMillis by 60000
+
+        log.info(s"Crunching ${flights.size} flights, ${minuteMillis.length} minutes (${SDate(crunchStartMillis).toISOString()} to ${SDate(crunchEndMillis).toISOString()})")
+        val startTime = System.currentTimeMillis()
+        val loads = portDeskRecs.flightsToLoads(FlightsWithSplits(flights), crunchStartMillis)
```

Comment:
it would be helpful to put types on these vals

Reply:
Done :+1: